[FEAT/#3] 번호 인증 기능 구현 - 인증 생성 기능

yummygyudon opened the pull request:
<!--
name: Makers Auth Feature PR template
about: 구현한 기능과 변경 사항에 대해 구체적으로 작성해주세요
title: '[FEAT/{#이슈번호}] 기능 내용'
labels: ''
assignees: ''
-->

<!--
- 리뷰어 추가했나요?
- 허가자 추가했나요?
- 라벨 추가했나요?
-->

## Related Issue 🚀
- related #3

모든 기능을 한 번에 PR할 경우, 너무 비대해질 것이 우려되어 상세 기능별로 PR을 하고자 합니다.

<br/>
<br/>

## Work Description ✏️
> 🚨 유의 : 프로젝트 초기 단계와 맞물려 다른 분들의 PR 내 코드들과 불일치하는 부분이 많습니다!! 리뷰 및 논의를 통한 최종 구조 확정 후에 확정 결과와 일치시키는 작업을 할 예정입니다. (ex. Rebase & Conflict Resolve + Merge)

- 번호 인증 기능 중 "번호 인증 생성" 기능을 구현했습니다. (자세한 기능 목록은 #3  이슈를 확인해주세요)

<br/>
<br/>

## PR Point 📸
<!-- 피드백을 받고 싶은 부분을 적어주세요. -->

### ※ BaseResponse
현재 구현된 BaseResponse의 경우, Response의 Body에 해당하는 객체로
API 컨트롤러 메서드/API 컨트롤러 Exception Handler 에서는 아래와 같이 반환해야 하는 상황입니다.
```java
return ResponseEntity
        .status({SuccessCode Enum}.getStatus.value())
        .body(BaseResponse.ofSuccess({SuccessCode Enum}, {Body Data Instance}));

// Exception Handler
return ResponseEntity
        .status({Listened Exception Instance}.getError().getStatus.value())
        .body(BaseResponse.ofFailure({Listened Exception Instance}.getError());
```

위 요소에 대해 고민되는 부분은
"기존 활용 방식을 유지할 것인지" 혹은 "ResponseEntity로 변환되는 Util 객체를 새롭게 구현하여 반복되는 코드양을 줄일 것인지" 입니다.

기존 방식의 경우, 모든 처리에서 Response Entity를 생성하는 코드가 반복되지만 Status Code와 Body에 어떤 데이터가 주입되는지 직관적으로 파악하기 용이하다는 장점이 있습니다.

@hyunw9 , @sung-silver  
두 분 의견 궁금합니다!!


<br/>

### ※ BaseEntity

Domain POJO 객체와 Database Entity 객체를 분리함으로서 고민해야 할 사항인 것 같은데요!!
우선 Domain 객체와 Entity가 활용되는 케이스는 두 가지 정도일 것 같습니다
- 기존에 저장되어 있던 데이터에 대한 Domain 객체를 수정하여 다시 Entity로 변환하고 저장하는 경우
  - Entity에 기존 식별 ID를 재사용할 수 있어야 합니다.
- 새롭게 생성되는 데이터에 대한 Domain 객체를 Entity로 변환하고 저장하는 경우
  - Entity에 새로운 식별 ID가 할당되어야 합니다.

이에 본 PR에서는
`BaseEntity.class` Abstract 클래스 내에 Setter 메서드를 구현하여 해결했습니다.
이 때, `protected`를 지정하여 상속받은 Entity 들만이 ID를 주입할 수 있도록 했습니다.


@hyunw9 , @sung-silver  
이 부분에 대해 다른 분들과 더 토론을 해보고 싶어서 가지고 왔습니다!!
다양한 의견과 우려스러운 점이 있다면 가감없이 전달해주시면 감사하겠습니다!

<br/>

### ※ Port - `In` & `Out`
우선 Port/Adapter 패턴 적용을 위해
아래와 같은 구조를 기반으로 구현했습니다.

> In => Out
```
(Client)
    ⬇️
Request DTO
    ⬇️
(Controller)
    ⬇️
Command
    ⬇️
(Service :: In Port Usecase Impl)
- Domain 객체 활용
    ⬇️
Domain -> Entity
    ⬇️
(Repository :: Out Port Usecase Impl)
```

> Out => In
```
(Repository :: Out Port Usecase Impl)
    ⬇️
Entity -> Domain
    ⬇️
(Service :: In Port Usecase Impl)
- Domain 객체 활용
    ⬇️
Info (VO) 👉 "미구현 상태"
    ⬇️
(Controller)
    ⬇️
Response DTO
    ⬇️
(Client)
```
위 구조를 통해 Domain 객체는 외부로부터 영향을 받지 않고
외부 변경에 의존적이지 않은 구조를 유지할 수 있게 된 것 같습니다!!

또한 외부(Controller)부터 전달받는 데이터를 Command 객체로 정의하여 주입 데이터를 표준화 했으며
관심사 분리를 위해 Command 객체는 실질적으로 사용하는 Usecase 인터페이스 내부에 구현하였습니다!

@hyunw9 , @sung-silver  
이에 추가로 `Info`라는 외부에 전달하는 VO 객체를 Command 객체와 동일하게 내부에 구현하는 것은 어떻게 생각하시나요?!
근거는 아래와 같구 여러분들의 의견이 궁금합니다.
- Domain 객체를 Controller에 바로 제공하는 것은 변경에 취약해질 수 있다. (Response - Domain 간 의존성이 형성됨)

<br/>

### ※ 패키징 
아직 구현되지 않아 가시적으로 패키지 구조가 표현되지 않은 상태여서
예시와 함께 제가 구상한 패키징 규칙을 제안드려봅니다!!

@hyunw9 , @sung-silver  
가감 없는 피드백과 의견 부탁드려요!!

> `application`

- "**도메인**"을 기준으로 2차 패키징을 진행합니다.  
- "**api**" / "**dto**"로 3차 패키징을 진행합니다.
```
application
ᄂ (Domain Name)
    ᄂ api
        ᄂ xxxApi (interface)
        ᄂ xxxApiController(class) implements xxxApi
    ᄂ dto
        ᄂ request
            ᄂ (Domain Name)Reqeust.class (final Class) : 실 DTO 객체는 내부 Record로 구현
                - (Action).class (Record) : UseCase에 맞는 Command 전환 메서드 구현
        ᄂ response
            ᄂ (Domain Name)Response.class (final Class) : 실 DTO 객체는 내부 Record로 구현
                - (Action).class (Record) : UseCase에 맞는 Command 전환 메서드 구현
```

<br/>

> `database`

- 1차 분리 기준 : 데이터베이스 종류
- 2차 분리 기준 : `entity`/`repository
  - entity : Entity 객체들을 관리합니다.
  - repository : Repository 구현체 및 
(추후에 Entity 객체 / Repository 객체가 많아진다면 "도메인" 기준으로 패키지 분리를 진행해도 괜찮을 것 같습니다)
```
database
ᄂ (Database Name)
    ᄂ BaseEntity.class
    ᄂ entity
        ᄂ xxxEntity.class
    ᄂ repository
        ᄂ xxxJpaRepository
        ᄂ xxxJdbcRepository
        ᄂ xxxRepository : Usecase Out Port 구현체
```

<br/>

> `domain`

- 1차 분리 기준 : 도메인
```
domain
ᄂ (ADomain)
    - Domain POJO Classes...
```

<br/>

> `external`

- 1차 분리 기준 : 외부 플랫폼 (ex. `gabia` / `aws` / ...)

```
external
ᄂ(External Platform Name)
    - Sub Feature Classes...
    - Main Feature Class : Usecase Out Port 구현체
```

<br/>

> `support`

- 부가 관심사에 따라 기준은 존재하지 않습니다 (아래 예시 참고해주세요!)
```
support
ᄂ code
   ᄂ A Domain
      ᄂ failure
      ᄂ success
ᄂ common
ᄂ config
   - xxxConfig.class (`@Configuration` 클래스)
ᄂ constant
   - xxxConstant.class (final 클래스)
ᄂ exception
   ᄂ database
   ᄂ domain
   ᄂ external
   ᄂ support
   ᄂ ApplicationExceptionHandler.class(`@RestControllerAdvice` 클래스 & `@ExceptionHandler` 메서드)
ᄂ jwt
   ᄂ provider
ᄂ security
   ᄂ authentication
   ᄂ filter
ᄂ util
   - xxxUtil.class (final 클래스)
ᄂ value
   - xxxProperty.class (Record 클래스)
```

<br/>

> `usecase`

- 1차 분리 기준 : 행위 관심사 (ex. `auth` / `message` / ...)
- 2차 분리 기준 : `port` / `service`
  - `port` : `in` / `out` 3차 분리 & 각 Usecase 인터페이스 정의(`Command` 객체 정의)
  - `service` : 각 Usecase 인터페이스 구현체 
```
usecase
ᄂ`Action Interest`
    ᄂ`port`
        ᄂ `in`
           - Service Usecase 인터페이스
        ᄂ `out`
           - Repository / External Feature 인터페이스 
    ᄂ `service`
        ᄂ xxxService : Usecase In Port 구현체
```

yummygyudon committed:
[CHORE] 잘못된 property 매핑 수정 (`external.yaml`)

- 잘못된 Gabia 값 매핑을 수정했습니다.

yummygyudon committed:
[STYLE] 패키지 구조 개편

- 관심사를 재정의하여 패키지 구조를 개편했습니다.

yummygyudon committed:
[FEAT](`support.exception`) External 관심사 내 외부 서버 통신 Client에서 발생하는 예외 관리하는 Error/Exception 구현

- Client 사용 과정에서 발생할 수 있는 예외에 대해 관리합니다.
- ClientException의 경우, ClientError만 주입받을 수 있습니다.

* 외부 서버 API 공식 문서를 참고하여 발생할 수 있는 ErrorCode를 미리 선언했습니다.

yummygyudon committed:
[CHORE] Gabia 공식 문서 기반 구현을 위한 의존성 추가(OkHttp, Gson)

yummygyudon committed:
[FEAT] Gabia 서버 통신 객체를 구현합니다.

"SMS 문자 발송"(`#sendSmsMessage`)과 "LMS 문자 발송"(`#sendLmsMessage`) 두 가지 API에 대한 통신 메서드를 구현했습니다.
- 각 발생 Exception 마다 try~catch를 통해 비즈니스 예외로 변환하여 던질 수 있도록 했습니다.

yummygyudon committed:
[FEAT] 외부에서 사용될 Gabia 서비스 객체 구현 및 커맨드 객체 구현

파라마터가 지나치게 많아지는 경우를 방지하고 Service 객체와의 협력을 위해서는 반드시 전용 Command Type을 사용하도록 강제할 필요가 있다고 판단해 외부 객체와의 협력 메시지로서 사용될 Command 객체를 구현했습니다.
- GabiaCommand 인터페이스를 슈퍼타입으로 활용하여 각 상세 Command 객체에서 구현하도록 하여 타입 안정성을 보장했습니다.
- Command와 함께 `MessageSendType`을 함께 주입해 Service객체 내에서 분기처리 할 수 있도록 했습니다.

yummygyudon committed:
[FEAT] 문자(메시지::`Message`) 도메인 객체 구현

문자 타입(SMS/LMS)을 표현하는 MessageType Enum 객체도 함께 구현했습니다.
- SMS/LMS 여부에 따른 정적 팩토리 생성자 메서드를 구현했습니다.

yummygyudon committed:
[FEAT] 문자 메시지 도메인에 대한 처리를 담당하는 Out Port 구현

- "문자 메시지를 전송한다"라는 유스케이스에 대해 `sendMessage` 메서드를 정의하고 파라미터로 `Message` 도메인만을 주입 받도록 규격화했습니다.

yummygyudon committed:
[FEAT] 문자 메시지 Out Port 구현 및 내 문자 전송 유스케이스 구현(`MessageSender`)

내부에선 GabiaClient를 주입 받아 사용하도록 했습니다.
- 주입된 `Message` 객체의 Type에 따라 SMS 전송/LMS 전송을 분기 호출하도록 했습니다.

yummygyudon committed:
[REFACTOR] GabiaClient 중복 코드 함수화

yummygyudon committed:
[CHORE] 사용하지 않는 불필요한 패키지 및 객체 삭제

yummygyudon committed:
[FEAT] 핸드폰 인증 도메인에 대한 데이터 처리를 담당하는 Out Port 구현

- 파라미터로 `PhoneVerification` 도메인 객체를 주입받도록 표준화했습니다.
- 반환 객체를 `PhoneVerification` 도메인 객체로 받도록 표준화했습니다.

yummygyudon committed:
[FEAT] 핸드폰 인증(`PhoneVerification`) & 핸드폰 인증 유형(`PhoneVerificationType`)도메인 객체 구현

생성자와 Builder에 대해 접근제어자를 Private으로 설정하여 제고하는 생성자 메서드를 활용하해 필요에 따라 생성되도록 했습니다.
- 최초 생성시 코드 주입 없이 새로운 코드를 생성하는 방식의 생성자 메서드`#create` 메서드를 구현했습니다.
- 기존에 존재하는 도메인 객체에 의해 코드 주입이 필요한 생성자 메서드`#of` 메서드를 구현했습니다.
- VerificationCode의 경우, 핸드폰 인증 도메인 내에서만 유효한 관심사이기 때문에 Inner Class로 선언하였으면 생성자 및 생성자 팩토리 메서드를 `private`으로 제한하여 외부에서의 생성을 방지했습니다.
  - 최초 생성시 `random()` 메서드를 호출하여 랜덤 코드를 발급하도록 했습니다.
  - 기존에 존재하는 도메인 객체에 의해 생성시 `of(String code)` 메서드를 호출하여 생성하도록 했습니다.

yummygyudon committed:
[FEAT] Base Entity 구현

- ID의 경우, DB에서 관리하는 식별값이기 때문에 Base entity에 포함시켜 해당 객체를 상속받는 Entity 객체들은 모두 ID를 자동으로 가질 수 있도록 했습니다.
- 기존에 존재하던 데이터에 대한 도메인 객체에서 Entity로 변환하는 경우에 새로운 ID가 생성되지 않도록 해야하기 때문에 `id`를 주입받는 생성자를 선언했고 `protected` 접근 제어자를 통해 상속받은 entity 내부에서만 사용가능하도록 제한했습니다.

* EntityListener 활성화를 위해 JpaConfig에 `@EnableJpaAuditing`을 추가했습니다.
* Entity 및 JpaRepository 객체를 인식할 수 있도록 `@EntityScan` & `@EnableJpaRepositories`또한 명시하여 설정했습니다.

yummygyudon committed:
[FEAT] PhoneVerification 엔티티 구현 및 Outport(`PhoneVerificationRespository`) 구현

- Domain 객체를 통해 Entity로 변환하는 static 메서드(`from`)를 선언했습니다.
  - 이미 Id가 할당되어 있는 상태였다면 새로운 ID가 생성되지 않도록 protected 메서드인 `setId()`를 통해 주입될 수 있도록 했습니다.
- querydsl 혹은 jdbc template이 필요한 요소가 없기에 기본적인 JPA Repository 객체를 구현했습니다.

* BaseEntity에 대해 id 주입 생성자가 아닌 Setter 메서드로 변경했습니다.

yummygyudon committed:
[FIX] PhoneVerification 도메인 객체의 식별값 필드 `id` 추가

- static 생성자에도 반영했습니다.

yummygyudon committed:
[FEAT] In port "번호 인증 생성" 유스케이스 구현

- `CreateVerificationCommand` Record 객체를 정의했습니다.
  - 본 유스케이스에서만 사용되는 관심사이기 때문에 내부 Record로 선언했습니다.

yummygyudon committed:
[FEAT] 번호 인증 생성 요청 DTO 구현

In port 내에 구현되어 있는 Command 객체로 변환하는 `toCommand()` 메서드를 선언했습니다.
생성 요청의 경우, 요청 이력을 식별하는 ID 값이 없기 때문에 별도 long 타입 데이터는 존재하지 않습니다.
- 인증 관심사에 대한 요청 객체 집합으로 묶어 일괄적으로 관리하기 위해 Super Class / Inner Class 형식으로 구현했습니다.
  -  상위 객체의 경우, 객체로 생성되면 안되기에 생성자를 private으로 제한했습니다.
  - 상위 객체의 경우, "실제 요청 DTO 객체를 관리"하는 Util 객체에 가까우며 상속이 되어선 안되기에 `final` 클래스로 선언했습니다.
  - 내부 객체인 "실 DTO" 객체는 필드값들의 불변성을 보장하기 위해 Record로 구현했습니다.

yummygyudon committed:
[FEAT] "번호 인증 생성" API 구현

- 성공에 대한 2XX 상태값의 Response 객체를 생성하기 위해 `AuthSuccess` Enum 객체를 구현했습니다.

yummygyudon committed:
[CHORE] 불필요한 예시 객체 제거

yummygyudon committed:
[FEAT] AuthException 에 대한 ExceptionHandler 구현

hyunw9 commented:
고봉 PR이군요 리뷰에 앞서 고생하셨습니다 !

리뷰 이전에 논의할 내용에 대한 제 생각입니다!

## BaseResponse

<현행 방식을 활용하는 것이 좋을 것 같다>
이유 : 반복되는 코드가 존재하는 것은 사실입니다. 하지만 그 경도가 치명적이지 않고, 상황에 유연하게 대처할 수 있는 범위 내라고 생각합니다. 
`Util/Helper`와 같은 클래스를 활용하면 더 직관적으로 표현이 가능하지만, 개발자가 파악해야 하는 깊이가 추가될 수 있다고 생각합니다. 
때문에 큰 문제가 없다면 현행을 유지해도 괜찮겠다는 생각이 드네요 ! 
(어느 정도 인증이 성숙된다면 공통 컴포넌트가 더 명확해 질 것이고, 그 때 일반화를 고려해봐도 좋을것 같습니다! ) 

## BaseEntity

<코드리뷰에 첨부>

JPA의 장점을 최대화 할 수 있는 구현 방법이라고 생각합니다 !
다만 궁금한 점이 조금 있어서 코드리뷰에 첨가하겠습니다! 

## Port - In & Out
<감싸는 것에 동의>
제가 이해하기로는 가장 핵심적인 부분은 의존성의 흐름이라고 이해했습니다. `Controller -> Domain` | `Domain Service -> Domain` 으로 흐르고 있기 때문에 문제는 없지만, Domain service에서 진행한 연산의 결과를 그대로 내려주는 상황만 존재하는 것이 아니기 때문에 중간 레이어를 통해 변환하는 것은 좋은 관행이라고 생각합니다. 


## 패키징 

좋은 것 같습니다 ! 다만 몇가지 네이밍 규칙 수정 제안드려봅니다 

1. `database`내부 패키지는 벤더보다 **core** , **secondary** 정도로 통일하는 것이 어떨까요 ? 
  @sung-silver 과는 한번 논의했던 이야기인데요, `psql`, `redis`와 같이 상세 스펙을 노출시키기 보다, 저희가 구분 가능할 정도로 분리하면 어떨까 싶은 생각이 듭니다 ! `core`, `secondary` 등등.. 
2. `database`단의 **xxxRepositoryImpl**(Port interface의 구현체)는 **xxxAdapter**의 접미사가 더 명확하지 않을까요 ? 
 이 부분도 사소합니다만 port를 구현하는 현재의 repositoryImpl은, `JPA Repository`를 주입받고 활용하여 usecase를 수행하는 어댑터의 역할을 수행한다고 생각합니다 ! 따라서 제안드려봅니다 ! ! !

hyunw9 commented on src/main/java/sopt/makers/authentication/domain/auth/PhoneVerification.java:
```diff
@@ -1,3 +1,70 @@
+@RequiredArgsConstructor(access = PRIVATE)
+public class PhoneVerification {
+
+  private final Long id;
```
`P1`
이 부분이 살짝 궁금했는데요, 저희는 도메인 모듈의 응집도를 위해 JPA entity와 분리해 모델을 구현하고 있는 걸로 알고 있습니다 ! 

제가 이해하기로는 @Id 는 Jpa Entity가 영속화 될때 Generation 전략에 따라 ID를 주입하는 것으로 알고 있습니다. 
따라서 Id는 database 레벨의 Unique함을 보장하기 위해 사용되는 컬럼이고, 따라서 POJO domain 보다, DB Entity 레벨에서 구현되어야 하는 것이지 않나! 라는 생각이 들었습니다. 

하지만 현재 방식은 Verification 생성시에 id를 generate해야 하는 구현 방식인 것 같아서 의견을 물어보고 싶었슴다!

yummygyudon replied:
저도 이부분이 굉장히 고민되었는데요!
그렇다면 기존 식별 Id가 존재하는 데이터를 운용한다고 했을 때, 운용 과정 중에 해당 도메인 객체의 식별 값은 어떻게 관리하는 것이 좋을까요?!!

예를 들어 존재하는 데이터(이미 ID가 할당되어 있는 도메인 데이터)에 대해 Update하는 기능을 구현한다고 할 때,
"Entity -> Domain" 전환하게 되면 Id 값은 모르게 될 것이고 update 처리 이후 `Domain -> Entity`로 변환하여 재저장을 하게된다면 
기존에 조회했던 Entity의 ID 값을 다시 사용해야 할 거에요!!

이 때의 **기존 Entity ID**는 어떻게 관리해야할지 고민입니다!!
Usecase 인터페이스에 선언된 메서드에 대한 구현 메서드 내에서 해당 Id 값을 가지고 있도록 하는 방식을 사용해야 할까요?!!
```java
//ex

public class xxxService implements xxxUsecase {

    private final xxxAdapter xxxAdapter; // Repository Out port 객체

    // UpdateResultInfo는 예제를 위한 임의의 VO 입니다.
    @Override
    public UpdateResultInfo update(UpdateCommand command) {
        final long id = command.id();
        XxxEntity fooEntity = xxxAdapter.findXxxById(id);
        Xxx fooDomain = fooEntity.toDomain();
        // update source codes
        ...

       // update되어 새롭게 생성된 Domain 객체
       Xxx updatedFooDomain = ....;
       XxxEntity updatedFooEntity =XxxEntity.from(updatedFooDomain);
       xxxAdapter.save(updatedFooEntity);
       return ...;
    }
}
```
이럴 경우, Entity 객체 내의 "Domain -> Entity" 변환 메서드인 `from()` 에 대해 2가지 케이스에 대해 오버로딩해서 구현해야겠네요!!

sung-silver replied:
저는 도메인에서도 id를 가지는게 맞다고 생각했습니다! 다만 현재 제 PR에서 이관한 도메인 객체의 id가 long으로 정의되어 있어 할당하지 않으면 0으로 초기화 될 것 같더라고요! Long으로 변경하겠습니다

yummygyudon replied:
> 결론

[SOPT makers 플랫폼 팀 BE 회의록](https://www.notion.so/sopt-makers/BE-fdf15d56593f4e7c89aa4d71c6d4291e?pvs=4) 참고

hyunw9 commented on src/main/java/sopt/makers/authentication/domain/auth/PhoneVerificationType.java:
```diff
@@ -0,0 +1,7 @@
+
+public enum PhoneVerificationType {
+  REGISTER,
+  SEARCH,
```
`P4`
SEARCH는 무엇을 표기하기 위한 Enum인지 궁금합니다!

yummygyudon replied:
가입 플랫폼 조회(Search)에 대한 값입니다!!!

가입 플랫폼 조회 시에도 번호 인증을 진행한다고 논의되었던 것으로 기억합니다!!

hyunw9 replied:
맞군요 확인입니다 !

hyunw9 commented on src/main/java/sopt/makers/authentication/external/gabia/GabiaClient.java:
```diff
@@ -0,0 +1,143 @@
+
+  private String extractSerializedDataIn(String dataKey, Response response) {
+    try {
+      HashMap<String, String> result =
+          new Gson().fromJson(Objects.requireNonNull(response.body()).string(), HashMap.class);
+      String data = result.get(dataKey);
+      if (data == null) {
+        throw new ClientResponseException(ClientError.GABIA_RESPONSE_UNAVAILABLE);
+      }
```
`P3`
이 부분 requireNonNull로 responseBody를 매핑하지만 하단부에서 널체크를 진행하고 있는 것 같은데,

body가 not null 인데 result(dataKey)의 데이터가 null인 상황이 존재하는지 궁금합니다 !

yummygyudon replied:
아하!! 요거는 문서를 확인하는 것이 가장 빠를 것 같은데요!!
[가비아 문자 서비스 API 개발안내서](https://message.gabia.com/api/documentation/)를 확인해보시면
성공 Response / 실패 Response의 형식과 필드명이 상이한 것을 확인할 수 있고
실패에서도 서로 다른 형식을 가지는 경우가 있습니다.

이에 원하는 키의 데이터(필드가 존재하지 않음::`result.get(dataKey)`)가 없을 경우, 예외를 발생시키도록 했습니다!!

hyunw9 commented on src/main/java/sopt/makers/authentication/external/gabia/GabiaClient.java:
```diff
@@ -0,0 +1,143 @@
+    headers.put("Content-Type", "application/x-www-form-urlencoded");
+    headers.put("Authorization", "Basic " + encodeAuthorization(authValue));
+    headers.put("cache-control", "no-cache");
+    return Headers.of(headers);
```
`P2`
헤더 부분도 위와 같이 상수화 시키면 좋을것 같습니다 !

hyunw9 commented on src/main/java/sopt/makers/authentication/support/code/external/failure/ClientError.java:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  @Override
+  public String getMessage() {
+    return additionalMessage == null
+        ? this.message
+        : String.format("%s (response message = %s)", this.message, this.additionalMessage);
+  }
```
`P3`
이 부분의 역할이 무엇인지 궁금합니다 !

yummygyudon replied:
외부 Client 로부터 발생한 에러일 경우,
외부 플랫폼에서 제공하는 에러 원인이 존재할 것입니다!!

이에 원인 파악 용이성을 위해 외부 플랫폼으로부터 수신한 에러 메시지까지 함께(`additionalMessage`) 보여줄 수 있도록 포맷팅하여 함께 보내주도록 하기 위함입니다!!

hyunw9 replied:
오 확인입니다

hyunw9 commented on src/main/java/sopt/makers/authentication/usecase/auth/port/out/PhoneVerificationRepository.java:
```diff
@@ -0,0 +1,7 @@
+
+import sopt.makers.authentication.domain.auth.PhoneVerification;
+
+public interface PhoneVerificationRepository {
```
`P5`
리얼 사소한데 여기 개행 안하면 성은이가 찾아옵니다 !

yummygyudon replied:
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 돔황쳐~~~~

sung-silver replied:
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 죄송해요 요즘.. 학교 수업이 클린코드라 안되어있으면 너무 잘보임 ㅜㅜ

hyunw9 commented on src/main/java/sopt/makers/authentication/usecase/auth/service/CreateVerificationService.java:
```diff
@@ -1,14 +1,42 @@
+    PhoneVerification phoneVerification =
+        PhoneVerification.create(command.name(), command.phone(), command.verificationType());
+
+    if (PhoneVerificationType.REGISTER.equals(phoneVerification.getVerificationType())) {}
```
`P4`
여기 부분 이해가 잘 안갑니다 ! 내부는 아직 구현이 안된걸까요 ???

yummygyudon replied:
어? 죄송합니다...

hyunw9 reviewed (changes requested):
마음이 풍부해지는 PR 고생하셨습니다! 

코드 살펴보며 몇가지 의문이 드는 내용 + 논의를 위해 적어주신 내용에 대해 작성해 보았씁니다 ! 

감사합니다 !

yummygyudon commented:
@hyunw9 
마음이 풍부해지셨다니 다행이네요ㅋㅋㅋㅋ

`패키징` 부분에 대한 의견에 저 또한 동의합니다!!
반영해보도록 하겠습니다!!!

`BaseResponse` 부분과 `Port - In & Out` 부분에 대해서는 @sung-silver 의 리뷰까지 받은 후에 반영해보도록 하겠습니다!

sung-silver commented on src/main/java/sopt/makers/authentication/application/auth/api/AuthApiController.java:
```diff
@@ -16,13 +15,15 @@
-    return null;
+      AuthRequest.CreatePhoneVerification createPhoneVerificationRequest) {
+    createVerificationUsecase.create(createPhoneVerificationRequest.toCommand());
+    return ResponseEntity.status(AuthSuccess.CREATE_PHONE_VERIFICATION.getStatus().value())
```
p4

`ResponseEntity.ok(BaseResponse.ofSuccess(AuthSuccess.CREATE_PHONE_VERIFICATION);`
으로 나타내는건 어떨까요?? 해당 요청에 대한 status가 무엇인지 AuthSuccess에 들어가서 확인해야하는 것 같아 명시적으로 나타낼 수 있는 방법 제안드려봅니다!

그리고 저는 성공 타입에 대한 것들은 static import를 자주 사용하는데요 다른분들의 컨벤션도 궁금하네요~ @hyunw9

yummygyudon replied:
> 결론

[SOPT makers 플랫폼 팀 BE 회의록](https://www.notion.so/sopt-makers/BE-fdf15d56593f4e7c89aa4d71c6d4291e?pvs=4) 참고

sung-silver commented on src/main/java/sopt/makers/authentication/database/BaseEntity.java:
```diff
@@ -0,0 +1,47 @@
+
+import java.time.LocalDateTime;
+
+import jakarta.persistence.*;
```
`P3`

와일드 카드 없애주시면 좋을 것 같습니다!

yummygyudon replied:
뜨헉!

sung-silver commented on src/main/java/sopt/makers/authentication/database/BaseEntity.java:
```diff
@@ -0,0 +1,47 @@
+  @Column(name = "updated_at")
+  private LocalDateTime updatedAt;
+
+  @Column(name = "deleted_at")
+  private LocalDateTime deletedAt;
```
`p5`

BaseEntity를 상속받는 모든 엔티티는 삭제된 시점을 기록하는 것일까요? 아니면 현재 인증서버여서 저희가 다루는 중심 엔티티가 유저에 대한 것이라 이렇게 표현한 것 인지도 궁금하네요..!

yummygyudon replied:
흠.. 사실 처음에는 유저의 탈퇴 이후에도 재가입 가능성을 고려해서 soft-delete를 위해 `deletedAt`을 추가하긴 했는데요!!
생각해보니 모든 엔티티에 `deletedAt`은 필요 없어서 `BaseEntity`에는 불필요하다고 생각되기도 하네요...

@hyunw9 , @sung-silver  두 분 의견 확인하고 반영하고 싶습니다!

sung-silver replied:
Soft Delete를 고려하시는 것 같아 User 엔티티에만 넣는게 좋을 것 같습니다!

yummygyudon replied:
@sung-silver 

우선 BaseEntity 내에서는 삭제했구
User Entity 내에는 추후 삭제 기능이 요구된다면 추가하도록하죠!!

sung-silver commented on src/main/java/sopt/makers/authentication/database/BaseEntity.java:
```diff
@@ -0,0 +1,47 @@
+public abstract class BaseEntity {
+
+  @Id
+  @Column(name = "id", nullable = false)
```
`p2`

`@Id` 어노테이션을 사용하면 pk를 지정하는 것이고, pk는 기본 옵션이 not null로 설정되는데 중복해서 nullable=false를 정의하지 않아도 될 것 같습니다!

sung-silver commented on src/main/java/sopt/makers/authentication/database/postgres/entity/PhoneVerificationEntity.java:
```diff
@@ -0,0 +1,50 @@
+import sopt.makers.authentication.domain.auth.PhoneVerification;
+import sopt.makers.authentication.domain.auth.PhoneVerificationType;
+
+import jakarta.persistence.*;
```
`p3`

와일드 카드!

sung-silver commented on src/main/java/sopt/makers/authentication/database/postgres/entity/PhoneVerificationEntity.java:
```diff
@@ -0,0 +1,50 @@
+
+  @Column(name = "type", nullable = false)
+  @Enumerated(EnumType.STRING)
+  PhoneVerificationType type;
```
`p2`

변수 선언 시 private 접근제어자 누락된 것 같습니다..!

sung-silver commented on src/main/java/sopt/makers/authentication/domain/auth/PhoneVerification.java:
```diff
@@ -1,3 +1,70 @@
 package sopt.makers.authentication.domain.auth;
 
-public class PhoneVerification {}
+import static lombok.AccessLevel.*;
```
`p3`

와일드카드!!

sung-silver commented on src/main/java/sopt/makers/authentication/database/postgres/entity/PhoneVerificationEntity.java:
```diff
@@ -0,0 +1,50 @@
+  public static PhoneVerificationEntity from(PhoneVerification phoneVerification) {
+    PhoneVerificationEntity phoneVerificationEntity =
+        new PhoneVerificationEntity(phoneVerification);
+    phoneVerificationEntity.setId(phoneVerification.getId());
```
`p5`

이건 저의 질문인데요 그럼 처음에 엔티티 생성 시점에는 getId()에 Null이 넘어가게 되고 자동으로 id 할당 -> 데이터베이스 저장 이후부터 id의 값이 할당됨

이 플로우 맞는걸까요?? (제가 그렇게 하고 있는데 제가 올린 PR에서는 BaseEntity에 id를 따로 빼지 않아서 확인하려고 여쭤봅니다!)

그리고 null일 수도 있는 부분에 대해서는 도메인 객체로부터는 Optional을 넘겨받고 Entity 생성 시에만 null 로 넘겨주는건 어떤지도 문의드립니다!

yummygyudon replied:
null 일 수도 있는 상황이 존재하기 때문에 (ex. 새롭게 생성되는 인증 요청 이력)
Domain 객체 내에 `Long` boxed 타입으로 선언했구

> 처음에 엔티티 생성 시점에는 getId()에 Null이 넘어가게 되고 자동으로 id 할당 -> 데이터베이스 저장 이후부터 id의 값이 할당됨

이해하신 바가 맞습니다!!

> null일 수도 있는 부분에 대해서는 도메인 객체로부터는 Optional을 넘겨받고

boxed long 타입 Wrapper 객체인 `Long`이기 때문에 Optional로 넘겨 받는 것이 맞습니다!!

sung-silver commented on src/main/java/sopt/makers/authentication/database/BaseEntity.java:
```diff
@@ -0,0 +1,47 @@
+  @Column(name = "deleted_at")
+  private LocalDateTime deletedAt;
+
+  protected void setId(Long id) {
```
이건 리뷰는 아니고 오늘 수업에서 배운 방식이라 반가워서..ㅎ

sung-silver commented on src/main/java/sopt/makers/authentication/database/BaseEntity.java:
```diff
@@ -0,0 +1,47 @@
+  @PreUpdate
+  public void preUpdate() {
+    this.updatedAt = LocalDateTime.now();
+  }
```
`p5`
`@CreatedDate`, `@LastModifiedDate`를 사용하지 않고 이런식으로 정의하는 게 어떤 차이가 있나요??
Spring Data Jpa를 사용하는만큼 어노테이션으로 처리해도 좋지 않을까..! 라는 생각이 들었습니다 사용하신 의도가 명확하다면 궁금해서 여쭤봅니다!!

yummygyudon replied:
오 한 번 써보고 싶었던 `@PrePersist`, `@PreUpdate` 여서 사용해봣는데
아무리 생각해봐도 성은이 의견처럼 직관적인 어노테이션을 사용하는게 좋겠네요!!

sung-silver commented on src/main/java/sopt/makers/authentication/domain/message/Message.java:
```diff
@@ -0,0 +1,33 @@
+  private final MessageType type;
+
+  public static Message sms(String receiver, String content) {
+    return Message.builder().receiver(receiver).content(content).type(MessageType.SMS).build();
```
`p5`
가독성을 위해 줄바꿈해주는건 어떨까요?

```
return Message.builder()
  .receiver(receiver)
  .content(content)
  .type(MessageType.SMS)
  .build();
```

yummygyudon replied:
아니 이거 개행해서 커밋했는데 자동으로 개행이 풀려요ㅠㅜㅠㅜ

sung-silver commented on src/main/java/sopt/makers/authentication/external/gabia/GabiaClient.java:
```diff
@@ -0,0 +1,143 @@
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
```
`p3`

근데 와일드 카드를 지양하는게 저희 서비스에서 정의한 클래스 사용시에만인가요? 반복되다보니 제가 잘못 이해하고 있었나싶기도해서... 다시 여쭤봅니도

yummygyudon replied:
아닙니다ㅋㅋㅋㅋ 성은이가 이해한게 맞구 제가 미처 짚지 못하고 넘어간 부분이에요!!
나이스 눈썰미~

sung-silver commented on src/main/java/sopt/makers/authentication/external/gabia/GabiaClient.java:
```diff
@@ -0,0 +1,143 @@
+    Response response = executeRequest(request);
+
+    String message = extractSerializedDataIn("message", response);
+    if (!message.trim().equals("Success")) {
```
`p3`

설명적 변수로 분리하면 좋을 것 같습니다!
`boolean isNotSuccess = !message.trim().equals("Success")`

그리고 Success 라는 문자열도 상수화하는건 어떨까요?

sung-silver commented on src/main/java/sopt/makers/authentication/external/gabia/GabiaClient.java:
```diff
@@ -0,0 +1,143 @@
+
+  protected void sendLmsMessage(String receiver, String title, String content) {
+    String authToken = authenticate();
+
```
`p4`
불필요한 줄바꿈인 것 같습니다..!

yummygyudon replied:
행위를 기준으로 구분지어 표현하고자 개행을 했는데 조금 불필요해보일까요?!

sung-silver commented on src/main/java/sopt/makers/authentication/external/gabia/GabiaClient.java:
```diff
@@ -0,0 +1,143 @@
+
+  protected void sendSmsMessage(String receiver, String content) {
+    String authToken = authenticate();
+
```
`p4`
불필요한 줄바꿈인 것 같습니다..!

sung-silver commented on src/main/java/sopt/makers/authentication/external/gabia/GabiaClient.java:
```diff
@@ -0,0 +1,143 @@
+    if (!message.trim().equals("Success")) {
+      ClientError error = ClientError.GABIA_RESPONSE_UNAVAILABLE;
+      error.addMessage(message);
+      throw new ClientRequestException(error);
```
`p3`

sendSmsMessage()에 정의된 부분이랑 유사한 부분이 많은 것 같은데, 값이 다른 부분은 파라미터로 분리하여 함수로 분리하는건 어떨까요?

sung-silver commented on src/main/java/sopt/makers/authentication/usecase/auth/service/CreateVerificationService.java:
```diff
@@ -1,14 +1,42 @@
+    PhoneVerification phoneVerification =
+        PhoneVerification.create(command.name(), command.phone(), command.verificationType());
+
+    if (PhoneVerificationType.REGISTER.equals(phoneVerification.getVerificationType())) {}
```
`p3`

if문 내에서 검증하는 식이 길어지게 되면 설명적 변수로 분리하면 좋을 것 같습니다!

yummygyudon replied:
옹 좋은 방법인 것 같습니다!!

sung-silver commented on src/main/java/sopt/makers/authentication/usecase/auth/service/CreateVerificationService.java:
```diff
@@ -1,14 +1,42 @@
+        PhoneVerification.create(command.name(), command.phone(), command.verificationType());
+
+    if (PhoneVerificationType.REGISTER.equals(phoneVerification.getVerificationType())) {}
+
```
`p3`
불필요한 개행인 것 같습니다!

sung-silver commented on src/main/java/sopt/makers/authentication/usecase/auth/service/CreateVerificationService.java:
```diff
@@ -1,14 +1,42 @@
+        Message.sms(
+            command.phone(),
+            convertCodeToMessage(phoneVerification.getVerificationCode().getCode()));
+
```
`p3`
불필요한 개행인 것 같습니다!

yummygyudon replied:
확인했습니다!!

단순 궁금한게 하나 있는데
spotless 때문인지 지멋대로 개행이 되던데...진짜 안예쁨...
이거 어쩌죠..ㅋㅋㅋㅋㅋㅋㅋ

sung-silver replied:
저희끼리 변수 선언과 관련된 개행에 대한 코멘트는 하지 않고 함수 본문 내에서 불필요한 개행에 대해 리뷰하자! 라는 규칙을 정하면 될 것 같습니다! 어떠신가요?!

sung-silver reviewed (changes requested):
다음은 PR point에 문의주신 내용에 대한 제 생각입니다!

1. BaseResponse
- 저는 기존 활용 방식 유지가 좋다고 생각했습니다! 아직 서비스 볼륨이 크지 않기 때문에 Util 객체를 구현하지 않더라도 괜찮을 것 같아요

2. BaseEntity
- 이전 제 엔티티 PR에서는 id필드는 포함시키지 않았었는데 지금 구현해주신 부분도 좋은 것 같습니다! 제가 포함하지 않았던 이유는 setter를 지양하고 싶어서였긴 했어요 ㅎ 다만 반복되는 코드가 엔티티마다 (set해주는 부분)이 추가되는 부분은 감안을 해야할 것 같네요

3. Info에 대하여
- 저도 좋다고 생각합니다 현재 패키징 및 구현 방식에서는 감싸는게 좋을 것 같아요

4. 패키징
- 현욱오빠랑 이야기한 내용을 바탕으로 현재는 core, cache로 구분되고 있는데요 카톡에서 논의했던 대로 rds, nosql도 좋을 것 같네요!

yummygyudon committed:
[STYLE] `database` 패키지 내 2차 분리 기준 반영

직접적으로 데이터베이스 플랫폼을 노출시키는 것을 지양하고자 했습니다.
- `rdb`(관계형 데이터베이스 플랫폼) / `nosql`(비-관계형 데이터베이스 플랫폼)
- [참고 PR](https://github.com/sopt-makers/sopt-auth-backend/pull/9)
  - [리뷰- @hyunw9](https://github.com/sopt-makers/sopt-auth-backend/pull/9#issuecomment-2472439405)
  - [리뷰- @sung-silver](https://github.com/sopt-makers/sopt-auth-backend/pull/9#pullrequestreview-2432812837)

yummygyudon committed:
[STYLE] 불필요한 개행 수정 및 Import 상세화

와일드 카드 Import되는 요소들에 대해 상세적으로 import 할 수 있도록 했습니다.

yummygyudon committed:
[CHORE] BaseEntity 생성일/수정일 필드 업데이트 방식 수정

- 기존 `@PrePersist`, `@PreUpdate` 사용 방식에서 더 직관적인 `@CreatedDate`, `@LastModifiedDate` 어노테이션으로 변경했습니다.
- 기존 `protected` Id 필드 Setter 메서드에 대해 Lombok을 통해 간소화했습니다.
- [관련 리뷰](https://github.com/sopt-makers/sopt-auth-backend/pull/9#discussion_r1840108040)

yummygyudon committed:
[STYLE] 가독성 개선을 위한 Builder 호출 로직 개행 (`Message#sms`)

yummygyudon commented:
다음 리뷰들([현욱 리뷰](https://github.com/sopt-makers/sopt-auth-backend/pull/9#discussion_r1839467592) / [성은 리뷰](https://github.com/sopt-makers/sopt-auth-backend/pull/9#discussion_r1840111815))에서는 서로 의견이 다른 부분이 있는 것 같은데요!!
어떻게 결정할까요?!!

저는 두 의견 모두 일리가 있다고 생각되어서요!!

- Domain 객체에 Id 필드가 있어야 한다.
- Domain 객체에 Id 필드는 필요 없다. (ID는 DBMS에서 식별값으로 사용되는 요소이기 때문)

yummygyudon committed:
[CHORE] `PhoneVerification` 엔티티 필드 접근제어자 누락 해결

yummygyudon committed:
[REFACTOR] `GabiaClient` 내 사용 절댓값 및 상수값 Constant 클래스 분리 & Response 검증 메서드 분리

hyunw9 commented:
> 다음 리뷰들([현욱 리뷰](https://github.com/sopt-makers/sopt-auth-backend/pull/9#discussion_r1839467592) / [성은 리뷰](https://github.com/sopt-makers/sopt-auth-backend/pull/9#discussion_r1840111815))에서는 서로 의견이 다른 부분이 있는 것 같은데요!! 어떻게 결정할까요?!!
> 
> 저는 두 의견 모두 일리가 있다고 생각되어서요!!
> 
> * Domain 객체에 Id 필드가 있어야 한다.
> * Domain 객체에 Id 필드는 필요 없다. (ID는 DBMS에서 식별값으로 사용되는 요소이기 때문)


어떻게 구현하는지는 저희가 정하는 부분이라고 생각합니다 ! 다만, Domain에는 어떠한 방식으로도 일관성이 있어야 한다고 생각해요 ! 
위에 동규형이 달아주신 구현 방법 확인해보았는데, 제 기준에서는 그 방식이 조금 더 적절하지 않나..? 라는 의견입니다. 
도메인객체가 ID를 달고 있는 순간부터 db에 종속된 구성이고, 결국 저희가 Domain - Entity를 분리한 목적성이 흐려진다고 생각했습니다 !

yummygyudon committed:
[CHORE] Branch Merge

yummygyudon committed:
[FIX] 공통 엔티티 내 불필요한 컬럼 필드(`deleted_at`) 제거

yummygyudon committed:
[FIX] Domain 내 Id 필드 제거 및 Entity 클래스 내 반영

오버로딩을 통해 id를 주입받는 경우와 받지 않는 경우에 대한 Entity 생성자 팩토리 메서드를 구현했습니다.

yummygyudon committed:
[CHORE] 추가 외부 라이브러리 버전 변수화

yummygyudon committed:
[CHORE] Test 소스 코드 내 Import 누락 추가
